hash/hash_comparator.go

```diff
@@ -558,3 +576,78 @@ func decodeMD5Hash(encodedHash string) (pf, salt, hash []byte, err error) {
 		return nil, nil, nil, ErrInvalidHash
 	}
 }
+
+// decodeLegacyHash decodes argon2id wrapped legacy password hash.
+// format: $legacy$alg=<md5|sha1|sha256|sha512>,format=<base64 encoded format, e.g: '{PEPPER}-{PWD}-{SALT}'>,i=<number of iterations>,(salt=<base64 encoded salt>,pepper=<base64 encoded pepper>)$<argon2id encoded hash>
+func decodeLegacyHash(encodedHash string) (phc, alg, format string, iterations int, salt, pepper []byte, err error) {
+	parts := strings.SplitN(encodedHash, "$", 4)
+	if len(parts) != 4 {
+		return "", "", "", 0, nil, nil, ErrInvalidHash
+	}
+	phc = "$" + parts[3]
+	legacyParams := strings.Split(parts[2], ",")
+	for _, lp := range legacyParams {
+		switch kv := strings.Split(lp, "="); kv[0] {
+		case "alg":
+			alg = kv[1]
```

Comment:
Here is a potential for index out of bounds panics, I'd propose to use this instead:

```suggestion
		k, v, found := strings.Cut(lp, "=")
		if !found {
			return "", "", "", 0, nil, nil, ErrInvalidHash
		}
		switch k {
		case "alg":
			alg = v
```